drgn_tools/task.py

```diff
@@ -185,6 +191,133 @@ def get_command(task: Object) -> str:
+    """
+
+    rss_file: int
+    rss_anon: int
+    rss_shmem: int
+
+    @property
+    def total(self) -> int:
+        return self.rss_file + self.rss_anon + self.rss_shmem
+
+
+def get_task_rss(task: Object, cache: Optional[dict] = None) -> TaskRss:
+    """
+    Return the task's resident set size (RSS) in pages
+
+    The task's RSS is the number of pages which are currently resident in
+    memory. The RSS values can be broken down into anonymous pages (not bound to
+    any file), file pages (those associated with memory mapped files), and
+    shared memory pages (those which aren't associated with on-disk files, but
+    belonging to shared memory mappings). This function returns a tuple
+    containing each category, but the common behavior is to use the "total"
+    value which sums them up.
+
+    :param task: ``struct task_struct *`` for which to compute RSS
+    :param cache: if provided, we can use this to cache the mapping of
+      "mm_struct" to RSS. This helps avoid re-computing the RSS value for
+      processes with many threads, but note that it could result in out of date
+      values on a live system.
+    :returns: the file, anon, and shmem page values
+    """
+    try:
+        mmptr = task.mm.value_()
+        if mmptr is None:
+            return TaskRss(0, 0, 0)
+    except AttributeError:
+        return TaskRss(0, 0, 0)
+    else:
+        if mmptr and cache and mmptr in cache:
+            return cache[mmptr]
+
+    prog = task.prog_
+
+    MM_FILEPAGES = prog.constant("MM_FILEPAGES").value_()
+    MM_ANONPAGES = prog.constant("MM_ANONPAGES").value_()
+    try:
+        MM_SHMEMPAGES = prog.constant("MM_SHMEMPAGES").value_()
+    except LookupError:
+        MM_SHMEMPAGES = -1
+
+    filerss = anonrss = shmemrss = 0
+
+    if mmptr:
+        filerss = task.mm.rss_stat.count[MM_FILEPAGES].counter.value_()
+        anonrss = task.mm.rss_stat.count[MM_ANONPAGES].counter.value_()
+        if MM_SHMEMPAGES >= 0:
+            shmemrss = task.mm.rss_stat.count[MM_SHMEMPAGES].counter.value_()
+
+    ltask = task.group_leader
+
+    filerss += ltask.rss_stat.count[MM_FILEPAGES].value_()
+    anonrss += ltask.rss_stat.count[MM_ANONPAGES].value_()
+    if MM_SHMEMPAGES >= 0:
+        shmemrss += ltask.rss_stat.count[MM_SHMEMPAGES].value_()
+
+    for gtask in list_for_each_entry(
+        "struct task_struct",
+        task.group_leader.thread_group.address_of_(),
+        "thread_group",
+    ):
+        filerss += gtask.rss_stat.count[MM_FILEPAGES].value_()
+        anonrss += gtask.rss_stat.count[MM_ANONPAGES].value_()
+        if MM_SHMEMPAGES >= 0:
+            shmemrss += gtask.rss_stat.count[MM_SHMEMPAGES].value_()
+    rss = TaskRss(filerss, anonrss, shmemrss)
+    if cache is not None:
+        cache[mmptr] = rss
+
+    return rss
+
+
+def get_rss(task: Object) -> int:
+    """
+    Return the task's resident set size (RSS) in kilobytes
+    """
+    prog = task.prog_
+    task_rss = get_task_rss(task).total
```

Comment:
Calling `get_task_rss(task)` without providing a cache parameter defeats the entire purpose of having it. Same goes for the call in `get_pct_mem()`. Only by providing the cache parameter can you ensure that the computation is efficient.

Honestly, I'd recommend deleting both helper functions and computing them directly in `show_taskinfo()`.